models/perm/access/repo_permission.go

```diff
@@ -18,6 +18,21 @@ import (
 	"code.gitea.io/gitea/modules/util"
 )
 
+type ErrPermissionDenied struct {
+	RepoID int64
+	Unit   unit.Type
+	Perm   perm_model.AccessMode
+}
+
+func (e ErrPermissionDenied) Error() string {
+	return fmt.Sprintf("permission denied to access repo %d unit %s with mode %s", e.RepoID, e.Unit.LogString(), e.Perm.LogString())
```

Comment:
It could use `Errorf("..., %w", util.ErrPermissionDenied)`

Then could check it by `errors.Is(err, util.ErrPermissionDenied)`

Reply:
Or remove this new error completely.

Just use `util.NewPermissionDeniedErrorf(".....")`

Reply:
cf1d71a8b6